ocaml/testsuite/tests/typing-zero-alloc/signatures.ml

```diff
@@ -0,0 +1,821 @@
+       is not included in
+         val f : int -> t [@@zero_alloc]
+       zero_alloc arity mismatch:
+       When using "zero_alloc" in a signature, the syntactic arity of
+       the implementation must match the function type in the interface.
+       Here the former is 2 and the latter is 1.
+|}]
+
+(********************************************************************)
+(* Test 7: A practicalish example of a non-obvious zero_alloc arity *)
+
+module type S7 = sig
+  val[@zero_alloc (arity 2)] f : int -> int -> int -> int*int
+end
+
+(* The expected behavior from the backend analysis for the two funtions below
+   is checked in [tests/backend/checkmach/test_arity.ml] *)
+
+module M7_1 : S7 = struct
+  let[@zero_alloc] f x y =
+    if x = y+1 then fun z -> (z,z) else fun z -> (z,0)
+end
+
+module M7_2 : S7 = struct
+  let[@zero_alloc] f x y z =
+    if x = y+1 then (z,z) else (z,0)
+end
+
+[%%expect{|
+module type S7 =
+  sig val f : int -> int -> int -> int * int [@@zero_alloc (arity 2)] end
+module M7_1 : S7
+Lines 13-16, characters 19-3:
+13 | ...................struct
+14 |   let[@zero_alloc] f x y z =
+15 |     if x = y+1 then (z,z) else (z,0)
+16 | end
+Error: Signature mismatch:
+       Modules do not match:
+         sig val f : int -> int -> int -> int * int [@@zero_alloc] end
+       is not included in
+         S7
+       Values do not match:
+         val f : int -> int -> int -> int * int [@@zero_alloc]
+       is not included in
+         val f : int -> int -> int -> int * int [@@zero_alloc (arity 2)]
+       zero_alloc arity mismatch:
+       When using "zero_alloc" in a signature, the syntactic arity of
+       the implementation must match the function type in the interface.
+       Here the former is 3 and the latter is 2.
+|}]
+
+(*************************************)
+(* Test 8: Parsing "(arity n)" works *)
+
+
+module type S8_1 = sig
+  val[@zero_alloc (arity 42)] f : int -> int
+end
+
+module type S8_2 = sig
+  val[@zero_alloc (arity 42) opt] f : int -> int
+end
+
+module type S8_3 = sig
+  val[@zero_alloc opt (arity 42)] f : int -> int
+end
+
+module type S8_4 = sig
+  val[@zero_alloc (arity 42) opt strict] f : int -> int
+end
+
+module type S8_5 = sig
+  val[@zero_alloc opt (arity 42) strict] f : int -> int
+end
+
+module type S8_6 = sig
+  val[@zero_alloc opt strict (arity 42)] f : int -> int
+end
+
+[%%expect{|
+module type S8_1 = sig val f : int -> int [@@zero_alloc (arity 42)] end
+module type S8_2 = sig val f : int -> int [@@zero_alloc opt (arity 42)] end
+module type S8_3 = sig val f : int -> int [@@zero_alloc opt (arity 42)] end
+module type S8_4 =
+  sig val f : int -> int [@@zero_alloc strict opt (arity 42)] end
+module type S8_5 =
+  sig val f : int -> int [@@zero_alloc strict opt (arity 42)] end
+module type S8_6 =
+  sig val f : int -> int [@@zero_alloc strict opt (arity 42)] end
+|}]
+
+(****************************************************)
+(* Test 9: (arity n) in structures gives warning 47 *)
```

Comment:
Why? It would be kind of nice to check the arity in the impl. Less difference between the language of zero_alloc for intfs vs. impls.

Reply:
In signatures, writing arity overrides the arity we would otherwise infer.  In structures, it could only be a check.  To me, it seemed potentially confusing to have it mean two different things in the two places, and therefore better to just disallow it.

Also, it feels odd to give users a way to check the arity of their funtions, but only if they are zero alloc.  (I already feel this way a bit about the signatures, but there at least there's a clear reason you might need it in combination with zero_alloc.)  Maybe we should just add a separate attribute for this, if we think it is useful?

Reply:
Fair enough. If we ever add such an attribute, we can remove `arity` from `zero_alloc` and have `zero_alloc` look for a separate `[@@arity n]` attribute instead, even in signatures.